Auction house to create auctions for fungible tokens

JackieXu opened the pull request:
Added a simple example of an auction house factory contract that can be called to create new auctions for fungible tokens.

JackieXu committed:
feat: add simple auction house

JackieXu committed:
chore: correct compiler version for solidity linter

JackieXu committed:
chore: remove Greeter example

JackieXu committed:
style: satisfy linter

sophismay commented on contracts/AuctionHouse.sol:
```diff
@@ -0,0 +1,125 @@
+    ) external {
+        //  liquidity checks
+        require(sellTokenAddress.balanceOf(msg.sender) >= amount, "AuctionHouse: Creator not liquid enough");
+        require(sellTokenAddress.allowance(msg.sender, address(this)) >= amount, "AuctionHouse: Not approved");
```
would it make sense to approve if allowance not enough? Or seller has to do this explicitly?

JackieXu replied:
> Or seller has to do this explicitly?

Correct. Someone else can't call the `approve` function of an ERC20 contract for you.  See [`approve`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L130) on the template ERC20 contract, and its internal [`_approve`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L281) function.

Technically you could implement another method on your ERC20 contract that allows others to approve for you, but that's outside the scope of this PR. 😬

sophismay commented on contracts/auctions/FungibleAuction.sol:
```diff
@@ -0,0 +1,177 @@
+    function isActive() public view returns (bool) {
+        return (
+            startTimestamp <= block.timestamp
+            && block.timestamp < startTimestamp + duration
```
startTimestamp.add(duration) ?

JackieXu replied:
In this case there is no need for a safe addition, because the constructor already guards against this overflow.

TSnark commented on contracts/AuctionHouse.sol:
```diff
@@ -0,0 +1,124 @@
+contract AuctionHouse {
+    /// @dev All fungible auctions that are either starting in the future, already active,
+    /// or haven't had their tokens withdrawn yet after finishing.
+    address[] public fungibleAuctions;
```
Consider swap this with an OZ enumerable set, it might save you some code

JackieXu replied:
Considered, and accepted. Might also be useful later on. Will change!

TSnark commented on contracts/auctions/FungibleAuction.sol:
```diff
@@ -0,0 +1,172 @@
+        // Safe subtraction. Already guarded in the constructor.
+        uint256 priceRange = maxPrice - minPrice;
+
+        // Safe subtraction. Already guarded via line 82.
```
Comments with reference to exact lines are a bit fragile

JackieXu replied:
Agreed. Will change!

TSnark commented on contracts/auctions/FungibleAuction.sol:
```diff
@@ -0,0 +1,172 @@
+/// @title Auction for fungible tokens
+contract FungibleAuction {
+    using SafeMath for uint256;
+
```
Why not using SafeERC20 for IERC20; and switching to safe transfers?

JackieXu replied:
Laziness. I agree that adding the `SafeERC20` lib for this is better. Also `IERC20` interface instead of the `ERC20` contract is better. Will change.

There used to be a check using `decimals` which isn't on the `IERC20` interface, hence why I changed it at some point. Or lack of coffee, who knows?

TSnark commented on contracts/auctions/FungibleAuction.sol:
```diff
@@ -0,0 +1,172 @@
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+// TODO: Switch to proper interface?
+interface IAuctionHouse {
```
Personal preference, external interface file and events defined in the interface as they are part of the interface

JackieXu replied:
I like your preference. Will move some stuff around to align.

TSnark commented on .solhint.json:
```diff
@@ -3,7 +3,7 @@
   "rules": {
     "code-complexity": ["error", 7],
-    "compiler-version": ["error", "^0.8.0"],
+    "compiler-version": ["error", "^0.7.3"],
```
why not 0.7.4?

JackieXu replied:
The rest of the Inverse repos use 0.7.3. Will keep it at 0.7.3 for convenience.

JackieXu committed:
chore: added use of util libs and interfaces

In addition, some shuffling around was done to better line up function
categories, i.e. state changing functions and read-only view functions.

Also added and/or changes a few comments to clarify usage of certain
"unsafe" operations.

JackieXu commented:
Now with better usage of utility libraries!

In addition, some shuffling around was done to better line up function categories, i.e. state changing functions and read-only view functions.

Also added and/or changed a few comments to clarify usage of certain "unsafe" operations.

JackieXu committed:
chore: use IERC20 instead of ERC20

JackieXu committed:
chore: simplify transfers

JackieXu committed:
test: add auction withdrawal test

crispspruce commented on contracts/AuctionHouse.sol:
```diff
@@ -0,0 +1,102 @@
+        uint256 tokensSold,
+        uint256 minPrice
+    ) external override {
+        // Acts as guard, reverts if msg.sender is not a contract created via
```
nit: you might wanna fix this comment.

JackieXu replied:
Some call it nitpicking, others call it a keen eye. I'm on it.

crispspruce commented on contracts/auctions/FungibleAuction.sol:
```diff
@@ -0,0 +1,165 @@
+        // Guard against timestamps *very* far into the future and/or long durations.
+        // TODO: Intended? Should an item that can no longer be sold return `minPrice`?
+        if (block.timestamp > startTimestamp.add(duration)) {
+            return minPrice;
```
How about returning lowestSellPrice instead? 

Its possible that minPrice is never even hit. In such a case, seems like lowestSellPrice might be a more meaningful number to return.

JackieXu replied:
I like the idea!

crispspruce commented on contracts/auctions/FungibleAuction.sol:
```diff
@@ -0,0 +1,165 @@
+        }
+
+        // The number of tokens required to buy the requested amount (or what is left of it).
+        uint256 buyPrice = getPriceInBuyTokens(amount);
```
nit: How about renaming getPrice to getRate, and inlining the multiplication of rate and amount here instead?

JackieXu replied:
"Rate", that's the word I was looking for! Will change!

JackieXu committed:
refactor: changed getPrice -> getRate

Minor changes:
 - inlined an internal function that was only used once.
 - fixed comment about function entry guarding using `EnumerableSet.remove`.

nourharidy commented on contracts/auctions/FungibleAuction.sol:
```diff
@@ -0,0 +1,159 @@
+    /// @dev Withdraw the tokens that were bid.
+    /// TODO: Allow auctioneer to withdraw tokens before auction is finished, in case of long durations?
+    function withdraw() external override {
+        require(msg.sender == auctioneer, "Auction: only auctioneer can withdraw");
```
In a case where governance is the auctioneer, a vote would be required to call this function. So it's better to let anyone call it instead

JackieXu replied:
Agreed! Removed.

nourharidy commented on contracts/auctions/FungibleAuction.sol:
```diff
@@ -0,0 +1,159 @@
+
+        uint256 availableAmount = sellTokenAddress.balanceOf(address(this));
+
+        require(availableAmount > 0, "Auction: sold out");
```
This is already checked in `isActive()`

JackieXu replied:
Good point! Removed.

JackieXu committed:
chore: allow anyone to call 'withdraw'

JackieXu committed:
chore: remove double balance check

JackieXu committed:
chore: remove double multiplication guard from auction construction

This is already guarded by the SafeMath library.

JackieXu committed:
test: add additional tests for AuctionHouse and FungibleAuction
